src/modules/rep.ts

```diff
@@ -5,158 +5,181 @@ import {
 		);
+		if (mentionUsers.size !== 1) return;
 
-		if (!ru) {
-			ru = await RepUser.create({ id: user.id }).save();
-		}
+		const recipient = mentionUsers.first()!;
 
-		return ru;
-	}
+		console.log('Creating a Rep with recipient', recipient);
 
-	@listener({ event: 'messageCreate' })
-	async onThank(msg: Message) {
-		const GIVE = '✅';
-		const PARTIAL_GIVE = '🤔';
-		const NO_GIVE = '❌';
-		const LAUGH = '😆';
+		await Rep.create({
+			messageId: msg.id,
+			channel: msg.channelId,
+			amount: 1,
+			recipient: recipient.id,
+			initialGiver: msg.author.id,
+			date: new Date().toISOString(),
+		}).save();
 
-		// Check for thanks messages
-		const isThanks = this.THANKS_REGEX.test(msg.content);
+		await msg.react(repEmoji);
+	}
 
-		if (msg.author.bot || !isThanks || !msg.guild) return;
+	@listener({ event: 'messageReactionAdd' })
+	async onRepReact(reaction: MessageReaction, user: User) {
+		console.log(reaction.emoji.id);
+		if (
+			!reaction.message.guild ||
+			user.id === this.client.user?.id ||
+			(reaction.emoji.id ?? reaction.emoji.name) !== repEmoji
+		)
+			return;
 
-		const allMentionUsers = msg.mentions.users;
-		const mentionUsers = allMentionUsers.filter(
-			user => user.id !== msg.member?.id,
-		);
-		if (mentionUsers.size < allMentionUsers.size) await msg.react(LAUGH);
-		if (!mentionUsers.size) return;
+		const msg = reaction.message;
+		const author = (await msg.fetch()).author;
 
-		const senderRU = await this.getOrMakeUser(msg.author);
-		// track how much rep the author has sent
-		// 3 possible outcomes: NO_GIVE, PARTIAL_GIVE and GAVE
-		let currentSent = await senderRU.sent();
+		console.log('Received rep reaction on', msg.id);
 
-		if (currentSent >= this.MAX_REP) return await msg.react(NO_GIVE);
+		if (user.id === author.id) {
+			console.log('Reacter is message author; removing reaction');
+			return removeReaction();
+		}
 
-		for (const user of mentionUsers.values()) {
-			if (currentSent >= this.MAX_REP)
-				return await msg.react(PARTIAL_GIVE);
+		console.log('Querying database for existing Rep');
 
-			// give rep
-			const targetRU = await this.getOrMakeUser(user);
+		let existingRep = await Rep.findOne({ messageId: msg.id });
 
-			await RepGive.create({
-				from: senderRU,
-				to: targetRU,
-			}).save();
+		if (existingRep) {
+			console.log('Found existing Rep', existingRep);
+			if (user.id === existingRep.recipient) {
+				console.log('User is recipient; removing reaction');
+				return removeReaction();
+			}
+			console.log('Existing amount is', existingRep.amount);
+			existingRep.amount++;
+			existingRep.save();
+			console.log('Incremented amount to', existingRep.amount);
+			return;
+		}
 
-			console.log('Gave one rep from', msg.author, 'to', user);
+		console.log('None found');
 
-			currentSent++;
+		let recipient = author.id;
+
+		if (recipient == this.client.user!.id) {
+			console.log('Recipient is bot; checking for message ownership');
+			let altRecipient = getMessageOwner(msg);
```

Comment:
I haven't really figured out the whole control flow since I'm doing other stuff ATM but this isn't persistent, is that okay?

Reply:
Yeah, on old messages, it won't work, but that's fine IMO as it's a bit of an edgecase.

Reply:
An edge case that could be used to get a lot of unremoveable rep?

Reply:
Once a message is used to give rep, the recipient is locked in, so it could still be removed

Reply:
Though I do need to remove rep from deleted messages, come to think of it